exist-core/src/main/java/org/exist/xquery/GeneralComparison.java

```diff
@@ -1040,6 +1041,48 @@ private AtomicValue convertForCompareAtomic(final AtomicValue value, final int t
         return value;
     }
 
+    private AtomicValue convertForValueComparison(final AtomicValue value, final int thisType, final int otherType)
+            throws XPathException {
+        /*
+            if the two operands are instances of different primitive types then:
+         */
+        if (Type.primitiveTypeOf(thisType) != Type.primitiveTypeOf(otherType)) {
+
+            /*
+                a. If each operand is an instance of one of the types xs:string or xs:anyURI,
+                   then both operands are cast to type xs:string.
+             */
+            if ((Type.subTypeOf(thisType, Type.STRING) || thisType == Type.ANY_URI)
+                    && (Type.subTypeOf(otherType, Type.STRING) || otherType ==Type.ANY_URI)) {
+                return value.convertTo(Type.STRING);
+            }
+
+            /*
+                b. If each operand is an instance of one of the types xs:decimal or xs:float,
+                then both operands are cast to type xs:float.
+             */
+            if ((Type.subTypeOf(thisType, Type.DECIMAL) || thisType == Type.FLOAT)
+                    && (Type.subTypeOf(otherType, Type.DECIMAL) || otherType == Type.FLOAT)) {
+                return value.convertTo(Type.FLOAT);
+            }
+
+            /*
+             *  c. If each operand is an instance of one of the types xs:decimal, xs:float, or xs:double,
+             *     then both operands are cast to type xs:double.
+             */
+            if ((Type.subTypeOf(thisType, Type.DECIMAL) || thisType == Type.FLOAT || thisType == Type.DOUBLE)
```

Comment:
Note to self: Test for behaviour with decimal subtypes and also subtypes of xs:integer